### PR TITLE
Multi kafka streams app support

### PR DESCRIPTION
**Pull request by surajpuvvada**
1. Added a container to run multiple KafkaStreamApp instances
2. Partitioned the RocksDBCacheProvider per application-id

### CODE REVIEW

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.typesafe.config.Config;
+import java.util.ArrayList;
+import java.util.List;
+import org.hypertrace.core.serviceframework.PlatformService;
+import org.hypertrace.core.serviceframework.config.ConfigClient;
+
+public abstract class MultiKafkaStreamInstanceApp extends PlatformService {
+
+  static final String SUB_TOPOLOGY_NAMES_CONFIG_KEY = "sub.topology.names";
```
Comment (kotharironak) [inline kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java:11]: Can we make this private?

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java [hunk 2]
```diff
@@ -54,38 +60,48 @@ public static RocksDBCacheProvider get() {
...
 
+  private CacheWrapper getCacheWrapper(String applicationId) {
+    return cacheConfigPerApplicationId.computeIfAbsent(applicationId,
+        s -> new CacheWrapper());
+  }
+
   protected synchronized void initCache(Options options, Map<String, Object> configs) {
-    if (cache == null) {
-      cacheTotalCapacity = DEFAULT_CACHE_TOTAL_CAPACITY;
+    String applicationId = (String) configs.get("application.id");
```
Comment (kotharironak) [inline kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java:69]: can we move `application.id` to private static var?

Comment (kotharironak) [inline kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java:69]: Can this `applicationId` be null?

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  private final List<KafkaStreamsApp> appsList = new ArrayList<>();
+
+  public MultiKafkaStreamInstanceApp(ConfigClient configClient) {
+    super(configClient);
+  }
+
+  @Override
+  protected void doInit() {
+    Config appConfig = getAppConfig();
+    List<String> apps = appConfig.getStringList(SUB_TOPOLOGY_NAMES_CONFIG_KEY);
```
Comment (laxmanchekka) [inline kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java:22]: these are not sub-topologies anymore. these are individual topologies in their respective streams applications.

`child.streams.apps` or `child.apps`?

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java [hunk 4]
```diff
@@ -32,16 +34,20 @@
...
 
   private static final RocksDBCacheProvider PROVIDER = new RocksDBCacheProvider();
 
-  private Cache cache;
-  private WriteBufferManager writeBufferManager;
-
-  private long cacheTotalCapacity;
-  private double writeBuffersRatio;
-  private double highPriorityPoolRatio;
+  private final ConcurrentMap<String, CacheWrapper> cacheConfigPerApplicationId = new ConcurrentHashMap<>();
```
Comment (laxmanchekka) [inline kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java:50]: `applicationCaches`

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java (hunk 2 above)
Comment (laxmanchekka) [reply kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java:69]: `group.id` is mandatory in kafka streams - can't be null.

File: kafka-streams-framework/src/test/java/org/hypertrace/core/kafkastreams/framework/rocksdb/BoundedMemoryConfigSetterTest.java [hunk 5]
```diff
@@ -41,6 +41,7 @@ public void setUp() {
     RocksDBCacheProvider.get().testDestroy();
     options = new Options();
     configs = new HashMap<>();
+    configs.put("application.id", "test-app");
```
Comment (laxmanchekka) [inline kafka-streams-framework/src/test/java/org/hypertrace/core/kafkastreams/framework/rocksdb/BoundedMemoryConfigSetterTest.java:44]: if possible, please add a separate test for multiple appliactaions

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package org.hypertrace.core.kafkastreams.framework;
+
+import com.typesafe.config.Config;
+import java.util.ArrayList;
+import java.util.List;
+import org.hypertrace.core.serviceframework.PlatformService;
+import org.hypertrace.core.serviceframework.config.ConfigClient;
+
+public abstract class MultiKafkaStreamInstanceApp extends PlatformService {
```
Comment (laxmanchekka) [inline kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java:9]:
```suggestion
public abstract class MultiKafkaStreamsApp extends PlatformService {
```

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java (hunk 1 above)
Comment (surajpuvvada) [reply kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java]: done

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java (hunk 2 above)
Comment (surajpuvvada) [reply kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java]: done

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java (hunk 3 above)
Comment (surajpuvvada) [reply kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java]: changed to streams.apps

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java (hunk 4 above)
Comment (surajpuvvada) [reply kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/rocksdb/RocksDBCacheProvider.java]: done

File: kafka-streams-framework/src/test/java/org/hypertrace/core/kafkastreams/framework/rocksdb/BoundedMemoryConfigSetterTest.java (hunk 5 above)
Comment (surajpuvvada) [reply kafka-streams-framework/src/test/java/org/hypertrace/core/kafkastreams/framework/rocksdb/BoundedMemoryConfigSetterTest.java]: parameterized the tests and tested for multiple application ids

File: kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java (hunk 6 above)
Comment (surajpuvvada) [reply kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/MultiKafkaStreamInstanceApp.java]: accepted

**Review by kotharironak — APPROVED** [review]

**Pull request merged**